### PR TITLE
Rework Context

### PR DESCRIPTION
**Pull request by tomaszdurka**
- Replace back CM_Log_Context_App with CM_Log_Context
- remove CM_Log_Context_App
- add CM_Log_Context_Extra
- store everything on Context
- adjust usages

### CODE REVIEW

**Comment by tomaszdurka** [comment]
@fvovan please review

Not sure in the end why I've put out `extra` into separate class. Probably can be reverted.

File: library/CM/ExceptionHandling/Handler/Abstract.php [hunk 1]
```diff
@@ -101,6 +101,8 @@ private function _getPrintSeverityMin() {
      */
     protected function _logException(Exception $exception) {
         $logLevel = CM_Log_Logger::exceptionSeverityToLevel($exception);
-        $this->getServiceManager()->getLogger()->addMessage('Application error', $logLevel, new CM_Log_Context_App(null, null, $exception));
+        $context = new CM_Log_Context();
```
Comment (fvovan) [inline library/CM/ExceptionHandling/Handler/Abstract.php]: I think new signatures became worse. imo passing ready context was much better.

File: library/CM/Log/Formatter/Text.php [hunk 2]
```diff
@@ -40,8 +40,9 @@ public function renderContext(CM_Log_Record $record) {
             }
             $data['extra'] = implode(', ', $extraText);
         }
-        if ($record->getContext()->getAppContext()->hasException()) {
-            $data['exception'] = $this->_renderException($record->getContext()->getAppContext()->getSerializableException());
+        if ($exception = $record->getContext()->getException()) {
+            $serializableException = new CM_ExceptionHandling_SerializableException($exception);
```
Comment (fvovan) [inline library/CM/Log/Formatter/Text.php]: What if we hide `new CM_ExceptionHandling_SerializableException()` inside context's method?

File: library/CM/Log/Context/Extra.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
+<?php
+
```
Comment (fvovan) [inline library/CM/Log/Context/Extra.php]: I think we need either to bring it back  to "just array representation" or to do something with `type`, if `extra` is a class now, `type` could be its attribute. But probably it's simpler just to revert and consider it as "flexible and easy data storage".

File: library/CM/ExceptionHandling/Handler/Abstract.php (hunk 1 above)
Comment (tomaszdurka) [inline library/CM/ExceptionHandling/Handler/Abstract.php]: I disagree. Having complicated constructor signature with a lot of nulls is much worse than simply using setter for each of them. It also makes code more readable. Also extracting something into variable helps read the code. Plus we don't have so many usages.

File: library/CM/Log/Formatter/Text.php (hunk 2 above)
Comment (tomaszdurka) [inline library/CM/Log/Formatter/Text.php]: I thought about that, but this is only used in some special cases where we want to store exception. From Context point of view we don't know it and simple getException does the job.

File: library/CM/Log/Context/Extra.php (hunk 3 above)
Comment (tomaszdurka) [inline library/CM/Log/Context/Extra.php]: Not sure about importance of `type`, but I agree that Extra class is something unneeded. It also works in a bit different way than other parts of Context - you need to dump it with `toArray`. I will revert it.

**Comment by tomaszdurka** [comment]
@fvovan please review

**Comment by fvovan** [comment]
lgtm

**Comment by tomaszdurka** [comment]
Adjusting sk in https://github.com/cargomedia/sk/issues/3360

**Comment by tomaszdurka** [comment]
@fvovan please confirm

**Comment by tomaszdurka** [comment]
Nothing changed - confirming on my own :)

**Pull request merged**